streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```
@@ -230,6 +230,8 @@ public void onPartitionsRevoked(final Collection<TopicPartition> assignment) {
                 // suspend active tasks
                 suspendTasksAndState();
             } catch (final Throwable t) {
+                // TODO: when this happens, we should still make sure the standby / active tasks list gets updated before returning
```

@guozhangwang is this not true already? In `suspendTasksAndState` we only throw an exception at the very end of the method.
Also, any tasks that failed to be suspended are closed with `task.close(false)`

---

The issue is, when one or more tasks get an error when suspending, we are still calling 

```
private void updateSuspendedTasks() {
        suspendedTasks.clear();
        suspendedTasks.putAll(activeTasks);
        suspendedStandbyTasks.putAll(standbyTasks);
    }
```

as the last step of the function call, and the failed tasks will still be added to the `suspendedTasks` and `suspendedStandbyTasks`, so if they get reassigned it may proceed to execute the resume procedure; and if they do not get reassigned the `close()` call will be executed again. Is that OK?

---

A more general question is that, if we ever got an exception in the `onPartitionRevoked` callback, we will just remember it and proceed as normal, and only re-throw it in the next `poll` call. I.e. we will still execute `onPartitionAssigned` as normal, which smells bad to me.

---

The exception also gets thrown. Won't this just cause the thread to die?

---

Any exception thrown in the `onXX` callback will be swallowed as a log entry by `AbstractCoordinator`, that is why we need to remember the exception and re-throw it later in the next poll call.